File: 回測模組/exporters/csv_exporter.py
"""
CSV 輸出模組
負責將回測結果輸出為 CSV 檔案
"""
import os
import logging
import pandas as pd
from typing import Dict, List
from strategy_modules.position_manager import TradeRecord

logger = logging.getLogger(__name__)

# 輸出路徑設定
OUTPUT_BASE_DIR = r"D:\回測結果"


class CSVExporter:
    """CSV 輸出器"""

    def __init__(self, output_base_dir: str = OUTPUT_BASE_DIR):
        """
        初始化 CSV 輸出器

        Args:
            output_base_dir: 輸出基礎目錄
        """
        self.output_base_dir = output_base_dir

    def export_summary_to_csv(self, results: List[Dict], date: str) -> None:
        """
        輸出統計結果到CSV

        Args:
            results: 統計結果列表
            date: 日期
        """
        if not results:
            logger.warning("沒有交易結果可輸出")
            return

        # 建立輸出目錄
        output_dir = os.path.join(self.output_base_dir, date)
        os.makedirs(output_dir, exist_ok=True)

        # 輸出檔案路徑
        csv_path = os.path.join(output_dir, f'backtest_summary_{date}.csv')

        # 寫入CSV
        df = pd.DataFrame(results)
        df.to_csv(csv_path, index=False, encoding='utf-8-sig')

        logger.info(f"\n統計結果已輸出至: {csv_path}")

        # 顯示總計
        total_pnl = df['損益'].sum()
        total_trades = df['進場次數'].sum()
        total_wins = sum(1 for _, row in df.iterrows() if row['損益'] > 0)

        logger.info("\n" + "=" * 60)
        logger.info(f"總計:")
        logger.info(f"  交易次數: {total_trades}")
        logger.info(f"  獲利次數: {total_wins}")
        if total_trades > 0:
            logger.info(f"  總勝率: {total_wins/total_trades*100:.2f}%")
        logger.info(f"  總損益: {total_pnl:,.0f}")
        logger.info("=" * 60)

    def export_detailed_trades_to_csv(self, all_trade_details: List[Dict], date: str) -> None:
        """
        輸出所有股票的詳細交易記錄到總表

        Args:
            all_trade_details: 所有股票的詳細交易記錄
            date: 日期
        """
        if not all_trade_details:
            logger.warning("沒有交易記錄可輸出")
            return

        # 建立輸出目錄
        output_dir = os.path.join(self.output_base_dir, date)
        os.makedirs(output_dir, exist_ok=True)

        # 輸出檔案路徑（使用原本的總表檔名）
        csv_path = os.path.join(output_dir, f'backtest_summary_{date}.csv')

        # 轉換為DataFrame並輸出
        df = pd.DataFrame(all_trade_details)

        # 依股票代碼和進場時間排序
        df = df.sort_values(['股票代碼', '進場時間', '出場批次'])

        # 輸出CSV
        df.to_csv(csv_path, index=False, encoding='utf-8-sig')

        logger.info(f"\n詳細交易記錄已輸出至: {csv_path}")

        # 顯示總計統計
        total_trades = df['交易編號'].groupby(df['股票代碼']).max().sum()
        total_pnl = df.groupby(['股票代碼', '交易編號'])['總損益金額'].first().sum()
        winning_trades = df.groupby(['股票代碼', '交易編號'])['總損益金額'].first()
        total_wins = sum(1 for pnl in winning_trades if pnl > 0)

        logger.info("\n" + "=" * 60)
        logger.info(f"總計:")
        logger.info(f"  股票數量: {df['股票代碼'].nunique()}")
        logger.info(f"  交易次數: {total_trades}")
        logger.info(f"  獲利次數: {total_wins}")
        if total_trades > 0:
            logger.info(f"  總勝率: {total_wins/total_trades*100:.2f}%")
        logger.info(f"  總損益: {total_pnl:,.0f}")
        logger.info("=" * 60)

        # 同時產生批次腳本需要的 backtest_results_{date}.csv 檔案（在專案根目錄）
        self._export_batch_script_csv(all_trade_details, date)

    def _export_batch_script_csv(self, all_trade_details: List[Dict], date: str) -> None:
        """
        產生批次腳本需要的 backtest_results_{date}.csv 檔案
        格式: trade_id,entry_time,stock_id,entry_price,exit_price,exit_ratio,exit_type

        重要：trade_id 不包含出場批次編號，以便 trade_pnl_analyzer.py 能正確
        將同一筆交易的多個出場批次分組，計算加權平均出場價格

        Args:
            all_trade_details: 所有股票的詳細交易記錄
            date: 日期
        """
        if not all_trade_details:
            return

        # 轉換為批次腳本需要的格式
        batch_records = []
        for detail in all_trade_details:
            record = {
                'trade_id': f"{detail['股票代碼']}_T{detail['交易編號']}",  # 不包含 _B{出場批次}
                'entry_time': detail['進場時間'],
                'stock_id': detail['股票代碼'],
                'entry_price': detail['進場價格'],
                'exit_price': detail['出場價格'],
                'exit_ratio': detail['出場比例'],
                'exit_type': detail['出場原因']
            }
            batch_records.append(record)

        # 輸出到專案根目錄
        batch_csv_path = f'backtest_results_{date}.csv'
        df = pd.DataFrame(batch_records)
        df.to_csv(batch_csv_path, index=False, encoding='utf-8-sig')

        logger.info(f"批次腳本CSV已輸出至: {batch_csv_path}")

File: CSharp/BacktestModule/Exporters/CsvExporter.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text;
using Microsoft.Extensions.Logging;
using BacktestModule.Core;
using BacktestModule.Strategy;

namespace BacktestModule.Exporters
{
    /// <summary>
    /// CSV exporter for backtest results.
    /// Exports summary statistics and detailed trade records to CSV files.
    /// </summary>
    public class CsvExporter : ICsvExporter
    {
        private readonly string _outputBaseDir;
        private readonly ILogger<CsvExporter> _logger;

        /// <summary>
        /// Default output base directory.
        /// </summary>
        public const string DefaultOutputBaseDir = @"D:\backtest_results";

        /// <summary>
        /// Initializes the CSV exporter.
        /// </summary>
        /// <param name="outputBaseDir">Output base directory path.</param>
        /// <param name="logger">Logger instance. Can be null.</param>
        public CsvExporter(
            string outputBaseDir = DefaultOutputBaseDir,
            ILogger<CsvExporter> logger = null)
        {
            _outputBaseDir = outputBaseDir;
            _logger = logger ?? Microsoft.Extensions.Logging.Abstractions.NullLogger<CsvExporter>.Instance;
        }

        /// <summary>
        /// Exports summary statistics to a CSV file.
        /// File path: {baseDir}/{date}/backtest_summary_{date}.csv
        /// </summary>
        /// <param name="results">List of per-stock statistics dictionaries.</param>
        /// <param name="date">Date string.</param>
        public void ExportSummaryToCsv(List<Dictionary<string, object>> results, string date)
        {
            if (results == null || results.Count == 0)
            {
                _logger.LogWarning("No trade results to export.");
                return;
            }

            // Create output directory
            string outputDir = Path.Combine(_outputBaseDir, date);
            Directory.CreateDirectory(outputDir);

            // Output file path
            string csvPath = Path.Combine(outputDir, $"backtest_summary_{date}.csv");

            // Write CSV with UTF-8 BOM for Excel compatibility
            WriteDictionaryListToCsv(results, csvPath);

            _logger.LogInformation("Summary statistics exported to: {CsvPath}", csvPath);

            // Show totals
            double totalPnl = results.Sum(r =>
                r.ContainsKey("\u640D\u76CA") ? Convert.ToDouble(r["\u640D\u76CA"]) : 0);
            int totalTrades = results.Sum(r =>
                r.ContainsKey("\u9032\u5834\u6B21\u6578") ? Convert.ToInt32(r["\u9032\u5834\u6B21\u6578"]) : 0);
            int totalWins = results.Count(r =>
                r.ContainsKey("\u640D\u76CA") && Convert.ToDouble(r["\u640D\u76CA"]) > 0);

            _logger.LogInformation("");
            _logger.LogInformation(new string('=', 60));
            _logger.LogInformation("Summary:");
            _logger.LogInformation("  Total trades: {TotalTrades}", totalTrades);
            _logger.LogInformation("  Winning stocks: {TotalWins}", totalWins);
            if (totalTrades > 0)
            {
                _logger.LogInformation("  Win rate: {WinRate:F2}%",
                    (double)totalWins / totalTrades * 100);
            }
            _logger.LogInformation("  Total PnL: {TotalPnl:N0}", totalPnl);
            _logger.LogInformation(new string('=', 60));
        }

        /// <summary>
        /// Exports detailed trade records (all stocks) to a summary CSV.
        /// Also generates a batch script CSV.
        /// </summary>
        /// <param name="allTradeDetails">All trade detail dictionaries across all stocks.</param>
        /// <param name="date">Date string.</param>
        public void ExportDetailedTradesToCsv(List<Dictionary<string, object>> allTradeDetails, string date)
        {
            if (allTradeDetails == null || allTradeDetails.Count == 0)
            {
                _logger.LogWarning("No trade records to export.");
                return;
            }

            // Create output directory
            string outputDir = Path.Combine(_outputBaseDir, date);
            Directory.CreateDirectory(outputDir);

            // Output file path (uses the same summary file name)
            string csvPath = Path.Combine(outputDir, $"backtest_summary_{date}.csv");

            // Sort by stock code, entry time, and exit batch
            var sorted = allTradeDetails
                .OrderBy(d => d.ContainsKey("\u80A1\u7968\u4EE3\u78BC") ? d["\u80A1\u7968\u4EE3\u78BC"]?.ToString() : "")
                .ThenBy(d => d.ContainsKey("\u9032\u5834\u6642\u9593") ? d["\u9032\u5834\u6642\u9593"]?.ToString() : "")
                .ThenBy(d => d.ContainsKey("\u51FA\u5834\u6279\u6B21") ? d["\u51FA\u5834\u6279\u6B21"]?.ToString() : "")
                .ToList();

            // Write CSV
            WriteDictionaryListToCsv(sorted, csvPath);

            _logger.LogInformation("Detailed trade records exported to: {CsvPath}", csvPath);

            // Show totals
            var grouped = sorted
                .Where(d => d.ContainsKey("\u80A1\u7968\u4EE3\u78BC") && d.ContainsKey("\u4EA4\u6613\u7DE8\u865F"))
                .GroupBy(d => new
                {
                    StockId = d["\u80A1\u7968\u4EE3\u78BC"]?.ToString(),
                    TradeNum = d["\u4EA4\u6613\u7DE8\u865F"]?.ToString()
                })
                .ToList();

            int totalTrades = grouped.Count;
            double totalPnl = grouped.Sum(g =>
            {
                var first = g.First();
                return first.ContainsKey("\u7E3D\u640D\u76CA\u91D1\u984D")
                    ? Convert.ToDouble(first["\u7E3D\u640D\u76CA\u91D1\u984D"])
                    : 0;
            });
            int totalWins = grouped.Count(g =>
            {
                var first = g.First();
                return first.ContainsKey("\u7E3D\u640D\u76CA\u91D1\u984D")
                    && Convert.ToDouble(first["\u7E3D\u640D\u76CA\u91D1\u984D"]) > 0;
            });

            int stockCount = sorted
                .Where(d => d.ContainsKey("\u80A1\u7968\u4EE3\u78BC"))
                .Select(d => d["\u80A1\u7968\u4EE3\u78BC"]?.ToString())
                .Distinct()
                .Count();

            _logger.LogInformation("");
            _logger.LogInformation(new string('=', 60));
            _logger.LogInformation("Summary:");
            _logger.LogInformation("  Stock count: {StockCount}", stockCount);
            _logger.LogInformation("  Total trades: {TotalTrades}", totalTrades);
            _logger.LogInformation("  Winning trades: {TotalWins}", totalWins);
            if (totalTrades > 0)
            {
                _logger.LogInformation("  Win rate: {WinRate:F2}%",
                    (double)totalWins / totalTrades * 100);
            }
            _logger.LogInformation("  Total PnL: {TotalPnl:N0}", totalPnl);
            _logger.LogInformation(new string('=', 60));

            // Also generate batch script CSV
            ExportBatchScriptCsv(allTradeDetails, date);
        }

        /// <summary>
        /// Generates the batch script CSV file for trade_pnl_analyzer.py compatibility.
        /// Format: trade_id, entry_time, stock_id, entry_price, exit_price, exit_ratio, exit_type
        /// trade_id does not include the exit batch number, so the analyzer can group
        /// multiple exit batches of the same trade to calculate weighted average exit price.
        /// </summary>
        /// <param name="allTradeDetails">All trade detail dictionaries.</param>
        /// <param name="date">Date string.</param>
        private void ExportBatchScriptCsv(List<Dictionary<string, object>> allTradeDetails, string date)
        {
            if (allTradeDetails == null || allTradeDetails.Count == 0)
                return;

            // Convert to batch script format
            var batchRecords = new List<Dictionary<string, object>>();

            foreach (var detail in allTradeDetails)
            {
                string stockCode = detail.ContainsKey("\u80A1\u7968\u4EE3\u78BC")
                    ? detail["\u80A1\u7968\u4EE3\u78BC"]?.ToString() : "";
                object tradeNum = detail.ContainsKey("\u4EA4\u6613\u7DE8\u865F")
                    ? detail["\u4EA4\u6613\u7DE8\u865F"] : 0;

                var record = new Dictionary<string, object>
                {
                    ["trade_id"] = $"{stockCode}_T{tradeNum}",  // Does not include _B{batch}
                    ["entry_time"] = detail.ContainsKey("\u9032\u5834\u6642\u9593") ? detail["\u9032\u5834\u6642\u9593"] : "",
                    ["stock_id"] = stockCode,
                    ["entry_price"] = detail.ContainsKey("\u9032\u5834\u50F9\u683C") ? detail["\u9032\u5834\u50F9\u683C"] : 0,
                    ["exit_price"] = detail.ContainsKey("\u51FA\u5834\u50F9\u683C") ? detail["\u51FA\u5834\u50F9\u683C"] : 0,
                    ["exit_ratio"] = detail.ContainsKey("\u51FA\u5834\u6BD4\u4F8B") ? detail["\u51FA\u5834\u6BD4\u4F8B"] : 0,
                    ["exit_type"] = detail.ContainsKey("\u51FA\u5834\u539F\u56E0") ? detail["\u51FA\u5834\u539F\u56E0"] : ""
                };
                batchRecords.Add(record);
            }

            // Output to project root directory
            string batchCsvPath = $"backtest_results_{date}.csv";
            WriteDictionaryListToCsv(batchRecords, batchCsvPath);

            _logger.LogInformation("Batch script CSV exported to: {CsvPath}", batchCsvPath);
        }

        /// <summary>
        /// Writes a list of dictionaries to a CSV file with UTF-8 BOM encoding.
        /// All keys from all dictionaries are collected as headers. Missing keys produce empty cells.
        /// </summary>
        /// <param name="records">List of dictionaries to write.</param>
        /// <param name="filePath">Output file path.</param>
        private void WriteDictionaryListToCsv(List<Dictionary<string, object>> records, string filePath)
        {
            if (records.Count == 0)
                return;

            // Collect all unique keys in order of first appearance
            var allKeys = new List<string>();
            var keySet = new HashSet<string>();
            foreach (var record in records)
            {
                foreach (var key in record.Keys)
                {
                    if (keySet.Add(key))
                    {
                        allKeys.Add(key);
                    }
                }
            }

            // Write with UTF-8 BOM for Excel to correctly display Chinese characters
            using var writer = new StreamWriter(filePath, false, new UTF8Encoding(true));

            // Write header
            writer.WriteLine(string.Join(",", allKeys.Select(EscapeCsvField)));

            // Write data rows
            foreach (var record in records)
            {
                var values = allKeys.Select(key =>
                {
                    if (record.TryGetValue(key, out object value) && value != null)
                    {
                        return EscapeCsvField(FormatValue(value));
                    }
                    return "";
                });
                writer.WriteLine(string.Join(",", values));
            }
        }

        /// <summary>
        /// Formats a value for CSV output. Handles DateTime, double, and other types.
        /// </summary>
        private static string FormatValue(object value)
        {
            if (value == null)
                return "";

            if (value is DateTime dt)
                return dt.ToString("yyyy-MM-dd HH:mm:ss.fff", CultureInfo.InvariantCulture);

            if (value is double d)
                return d.ToString(CultureInfo.InvariantCulture);

            if (value is float f)
                return f.ToString(CultureInfo.InvariantCulture);

            if (value is decimal dec)
                return dec.ToString(CultureInfo.InvariantCulture);

            return value.ToString();
        }

        /// <summary>
        /// Escapes a CSV field value by quoting it if it contains commas, quotes, or newlines.
        /// </summary>
        private static string EscapeCsvField(string field)
        {
            if (string.IsNullOrEmpty(field))
                return "";

            if (field.Contains(',') || field.Contains('"') || field.Contains('\n') || field.Contains('\r'))
            {
                return "\"" + field.Replace("\"", "\"\"") + "\"";
            }
            return field;
        }
    }
}
